### PR TITLE
Add Job Badge Action Test

### PR DESCRIPTION
**Pull request by yashpal2104**
Improve coverage the JobBadgeAction class and the StatusImage class.

### Testing done

Confirmed automated tests passed and coverage increased.

### Submitter checklist
- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did

### CODE REVIEW

**Comment by yashpal2104** [comment]
@MarkEWaite I have made a draft PR please review these changes

**Comment by yashpal2104** [comment]
Hey @MarkEWaite any feedback on this would really help so that next time I would be sure of my work and make no errors. Also anything you would expect me to know what I did bad

**Review by MarkEWaite — COMMENTED** [review]
Comments are attached to various locations in the code.

File: src/test/java/org/jenkinsci/plugins/badge/actions/JobBadgeActionTest.java [hunk 1]
```diff
@@ -59,6 +71,82 @@ public static void createAction() throws Exception {
...
+
+        // Assert that the response contains the expected badge content
+        if (matcher.find()) {
+            String extractedText = matcher.group(1);
+            System.out.println("Extracted text: " + extractedText);
+            // Assert that the response contains the expected badge content
+            assertThat(response, containsString("passing"));
+        } else {
+            System.out.println("No match found");
+            assertThat("No match found", is("No match found"));
```
[the remark this answers is not in the document]
Comment (yashpal2104) [reply src/test/java/org/jenkinsci/plugins/badge/actions/JobBadgeActionTest.java:109]: Yeah this it was for my own reference I know this doesn't match with anything in the page, so I was going to ignore this I used it just to see if I could turn this into something which gets triggered when a response that does not hit the 'if' statement.

File: src/test/java/org/jenkinsci/plugins/badge/actions/JobBadgeActionTest.java [hunk 2]
```diff
@@ -59,6 +71,82 @@ public static void createAction() throws Exception {
...
 
+    @Test
+    public void setUpActionToBadgeBuildAndRun() throws IOException {
+        // Create an instance of JobBadgeAction
+        jobBadgeActionBuildAndRun = new JobBadgeAction(successfulAction.project);
+
+        // Get the Jenkins URL
+        jenkinsUrl = j.getURL().toString();
+        badgeUrl = jenkinsUrl + "job/" + jobBadgeActionBuildAndRun.getUrlEncodedFullName()
+                + "/build/badge/icon?build=1&style=style&subject=subject&status=status&color=green&config=config&animatedOverlayColor=animatedOverlayColor&link=link";
```
[the remark this answers is not in the document]
Comment (yashpal2104) [reply src/test/java/org/jenkinsci/plugins/badge/actions/JobBadgeActionTest.java:82]: There is a typo from my end I was going to test with the plain protected link which has 'job/' and one with the unprotected link 'buildStatus/icon' so yeah that's on me, thank you for pointing this out

File: src/test/java/org/jenkinsci/plugins/badge/actions/JobBadgeActionTest.java [hunk 3]
```diff
@@ -59,6 +71,82 @@ public static void createAction() throws Exception {
...
+
+        // Open a connection to the badge URL
+        URL url = new URL(badgeUrl);
+        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+        connection.setRequestMethod("GET");
+
+        // Read the response from the connection
+        Scanner scanner = new Scanner(connection.getInputStream());
+        scanner.useDelimiter("\\A");
+        response = scanner.hasNext() ? scanner.next() : "";
```
[the remark this answers is not in the document]
Comment (yashpal2104) [reply src/test/java/org/jenkinsci/plugins/badge/actions/JobBadgeActionTest.java:93]: Yes I will check out this approach although previously I did find this approach and tried it but I was not successful in it I will try it once again

**Comment by yashpal2104** [comment]
@MarkEWaite how does this look although I was only able to increase the line coverage to 69% and branch coverage to only 35%, should I keep improving until it reaches 80% ?

File: src/test/java/org/jenkinsci/plugins/badge/actions/JobBadgeActionTest.java [hunk 4]
```diff
@@ -59,45 +70,84 @@ public static void createAction() throws Exception {
...
-        assertThat(notBuiltAction.getDisplayName(), is("Embeddable Build Status"));
-        assertThat(successfulAction.getDisplayName(), is("Embeddable Build Status"));
-    }
+        // Get the Jenkins URL
+        jenkinsUrl = j.getURL().toString() + "job/" + jobBadgeActionBuildAndRun.getUrlEncodedFullName() + "/" + "badge/icon";
+        badgeUrl = jenkinsUrl + "?build=1&style=style&subject=subject&status=status&color=green&config=config&animatedOverlayColor=animatedOverlayColor&link=link";
+        // Open a connection to the badge URL
+        URL url = new URL(badgeUrl);
+        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+        connection.setRequestMethod("GET");
```
Comment (MarkEWaite) [inline src/test/java/org/jenkinsci/plugins/badge/actions/JobBadgeActionTest.java]: I don't understand how this connection is used.  As far as I can tell from the [javadoc](https://javadoc.scijava.org/Java11/java.base/java/net/HttpURLConnection.html), this must be unused

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
I've added comments that may be helpful and have pushed changes that contain the changes and leave a failing test to show you where more work is needed.

**Comment by yashpal2104** [comment]
Sorry I was in the middle of my college classes I will get on it and make the requested changes asap

**Comment by yashpal2104** [comment]
I am very close to a solution for the JobBadgeActionTests, I just need to learn how do I make my own customisable badge so that I can implement the learnings and implement that on the tests which will increase the coverage, I also then want to replicate the url in the tests so that the URL points to the correct page.
@MarkEWaite any help would be appreciated.

**Comment by yashpal2104** [comment]
Hey @MarkEWaite I have implemented a solution that I think would increase the coverage, I have added the comments where I think I am lacking and need a little bit of refinement. Could you take a look and let me know if there are any adjustments or corrections I should make? Your thoughts on handling the JSON parsing would be especially helpful and what mistakes I am making.

**Comment by yashpal2104** [comment]
Thank you @MarkEWaite for applying the changes, sadly I was not able to be of much help to the changes any feedback would be helpful for me to try and improve from my side.

**Comment by MarkEWaite** [comment]
> any feedback would be helpful for me to try and improve from my side.

You did good work, especially for a first effort.  I ran the tests in the debugger with breakpoints in the target class and then made adaptations to the tests to assure both test coverage and test reliability.  When I'm on a quest to improve coverage, an IDE with breakpoints and a debugger are helpful tools.

**Review by MarkEWaite — APPROVED** [review]

**Comment by yashpal2104** [comment]
> You did good work, especially for a first effort. I ran the tests in the debugger with breakpoints in the target class and then made adaptations to the tests to assure both test coverage and test reliability. When I'm on a quest to improve coverage, an IDE with breakpoints and a debugger are helpful tools.

Thank you for your kind words, I will learn more about the debugger and breakpoints tools and techniques to be able to make more meaningful contributions that would benefit the project. Thanks for your support!

**Pull request merged**